src/lib/raft/raft.c

```diff
@@ -729,10 +729,20 @@ raft_worker_handle_io(struct raft *raft)
 				 * to become a leader.
 				 */
 				raft_sm_wait_election_end(raft);
-			} else {
+			} else if (raft->last_demoted_term <
+				   raft->volatile_term) {
 				/* No leaders, no votes. */
 				raft_sm_schedule_new_vote(raft, raft->self,
 							  raft->vclock);
+			} else {
+				assert(raft->last_demoted_term ==
+				       raft->volatile_term);
+				/*
+				 * Resigned from the leader role in the current
+				 * term, hence, cannot become a candidate in the
+				 * current elections, so just wait.
+				 */
+				raft_sm_wait_election_end(raft);
```

Comment:
Nit: we already do the same logic just one if above, in `raft->vote != 0`. IMHO, it'll be easier to read, if this chunk will also go there with `if (raft->vote != 0 || raft->last_demoted_term == raft->volatile_term)`

Reply:
Fixed.